### PR TITLE
fix issue 18801 - std.stdio.File doesn't work with MSVCRT's UTF-8 mode

### PR DESCRIPTION
**Pull request by aG0aep6G**

### CODE REVIEW

File: std/stdio.d [hunk 1]
```diff
@@ -2835,13 +2838,32 @@ is empty, throws an `Exception`. In case of an I/O error throws
...
+
+            version (MICROSOFT_STDIO)
+            {
+                // Microsoft doesn't implement fwide. Instead, there's the
+                // concept of ANSI/UNICODE mode. fputc doesn't work in UNICODE
+                // mode; fputwc has to be used. So that essentially means
+                // "wide-oriented" for us.
+                immutable int mode = _setmode(f.fileno, _O_TEXT);
+                    // Set some arbitrary mode to obtain the previous one.
+                _setmode(f.fileno, mode); // Restore previous mode.
```
Comment (aG0aep6G) [inline std/stdio.d]: This is a bit awkward, but I couldn't find a function that just gets the mode.

**Comment by aG0aep6G** [comment]
I can't make sense of this.

The Win64 tester apparently sees the mode as `_O_WTEXT` when it should be `_O_U8TEXT`. But then it writes a UTF-8 BOM anyway. BUT THEN it encodes "foo" in UTF-16. And all that happens without D's File or LockingTextWriter being involved at all. It's just `fopen` and `fputwc`.

I'm open to suggestions.

**Comment by wilzbach** [comment]
(rebased)

**Comment by aG0aep6G** [comment]
Rebased properly (hopefully). I fully expect this to still fail in the same nonsensical way, though.

**Comment by aG0aep6G** [comment]
Looks like the new Windows testing setup turned this green. With the nonsensical failures gone, this might be good to go.

There is a failure on buildkite, though: Apparently, CyberShadow/ae segfaults. I'm not sure how to address that. Is it caused by this PR? Do I need to investigate it? @CyberShadow?

**Comment by Geod24** [comment]
https://github.com/CyberShadow/ae/issues/46

**Comment by aG0aep6G** [comment]
Sweet. All green.

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**